content/articles/rust-for-web-services.md

```diff
@@ -0,0 +1,132 @@
+This is not going to be exhaustive by any means but here are the things I liked and disliked the most.
+
+
+### Result and Option types
+Every Go developers has been caught forgetting the `if err != nil` when calling a function, every Python developers has seen exceptions that they didn't realise could happen. Rust (and some functional languages it's borrowing the concept from) bake those in the language, using the [Result](https://doc.rust-lang.org/std/result/) type.
+In practice, this means that by returning this type, the caller **has** to handle the possible errors. It wouldn't even compile if you didn't as we will see later, unless you use `.unwrap()` which is a way to get panics if there is an error.
+
+The Option types is the same as the Maybe in FP, it represents the possibility of a value which Javascript would benefit greatly from for example (`undefined is not a function` anyone?).
+
+
+### Pattern matching
+The easiest way to handle the above types is called *pattern matching* and is awesome in Rust.
+As an example from the demo, here's how to handle the Result from `list_passwords` that fetches all the passwords from the table in our handler:
+
+```rust
+match dal::list_passwords(conn) {
+    Ok(passwords) => {
+        let response_payload = try_or_500!(serde_json::to_string(&passwords));
+        Ok(Response::with((status::Ok, response_payload)))
+    },
+    Err(e) => {
+        Ok(Response::with((status::InternalServerError)))
+    }
+}
+```
+> **Note:** Rust automatically returns the last value, in that case the Ok() ones. Iron handler return type is a Result so I should probably return an Err but it currently works well enough™
+
+Removing the Err branch errors with the following message `error: non-exhaustive patterns: Err(_) not covered [E0004]`.
+
+Pattern matching is obviously not limited to Result and Option and can be used anywhere you would want a if/else or a switch.
+
+### Macros
+Macros are a way to abstract some code but, instead of a function call, the macro is expanded where it's called at compilation time. In practice it means you can simplify code without the overhead of calling a function. The most used macro is probably [try!](https://doc.rust-lang.org/std/macro.try!.html) which returns the error if there is one or gives the result value otherwise. Here's an example from the demo: while trying things out I ended up with the following two lines in each http handler to get a connection to postgres.
+
+```rust
+let pool = req.get::<PRead<db::PostgresDB>>().unwrap();
+let conn = pool.get().unwrap();
+```
+As you can see, I use `unwrap` which means it can panic in case of errors. I wrote the following macro to clean it up:
+
+```rust
+// db.rs
+// Gets a connection from the pool from the given request or returns a 500
+macro_rules! get_pg_connection {
+    ($req:expr) => (match $req.get::<persistent::Read<db::PostgresDB>>() {
```

Comment:
why ? you would need to use match everywhere you need a connection as it would return a Result